src/Hosting.Services/OmexStatefulService.cs

```diff
@@ -6,41 +6,57 @@
-			m_serviceRegistrator = serviceRegistrator;
-		}
-
-		/// <inheritdoc />
-		protected override Task OnOpenAsync(ReplicaOpenMode openMode, CancellationToken cancellationToken)
-		{
-			m_serviceRegistrator.PartitionAccessor.SetValue(Partition);
-			return base.OnOpenAsync(openMode, cancellationToken);
-		}
-
-		/// <inheritdoc />
-		protected override IEnumerable<ServiceReplicaListener> CreateServiceReplicaListeners() =>
-			m_serviceRegistrator.ListenerBuilders.Select(b => new ServiceReplicaListener(c => b.Build(this), b.Name));
-
-		/// <inheritdoc />
-		protected override Task RunAsync(CancellationToken cancellationToken) =>
-			Task.WhenAll(m_serviceRegistrator.ServiceActions.Select(r => r.RunAsync(this, cancellationToken)));
-	}
+    /// <summary>
+    /// Omex implementation of stateful service fabric service
+    /// </summary>
+    public sealed class OmexStatefulService : StatefulService, IServiceFabricService<StatefulServiceContext>
+    {
+        private readonly OmexStatefulServiceRegistrator m_serviceRegistrator;
+        private readonly IAccessor<ReplicaRoleWrapper> m_replicaRoleAccessor;
+
+        internal OmexStatefulService(
+            OmexStatefulServiceRegistrator serviceRegistrator,
+            StatefulServiceContext serviceContext)
+                : base(serviceContext)
+        {
+            serviceRegistrator.ContextAccessor.SetValue(Context);
+            serviceRegistrator.StateAccessor.SetValue(StateManager);
+            m_serviceRegistrator = serviceRegistrator;
+            m_replicaRoleAccessor = (IAccessor<ReplicaRoleWrapper>)serviceRegistrator.RoleAccessor;
+        }
+
+        /// <inheritdoc />
+        protected override Task OnOpenAsync(ReplicaOpenMode openMode, CancellationToken cancellationToken)
+        {
+            m_serviceRegistrator.PartitionAccessor.SetValue(Partition);
+            return base.OnOpenAsync(openMode, cancellationToken);
+        }
+
+        /// <inheritdoc/>
+        protected override Task OnChangeRoleAsync(ReplicaRole newRole, CancellationToken cancellationToken)
+        {
+            m_serviceRegistrator.RoleAccessor.SetValue(new ReplicaRoleWrapper { Role = newRole });
+            return base.OnChangeRoleAsync(newRole, cancellationToken);
+        }
+
+        /// <inheritdoc />
+        protected override IEnumerable<ServiceReplicaListener> CreateServiceReplicaListeners() =>
+            m_serviceRegistrator.ListenerBuilders.Select(b => new ServiceReplicaListener(c => b.Build(this), b.Name));
+
+        /// <inheritdoc />
+        protected override Task RunAsync(CancellationToken cancellationToken) =>
+            Task.WhenAll(m_serviceRegistrator.ServiceActions.Select(r => r.RunAsync(this, cancellationToken)));
+
+        /// <summary>
+        /// Gets the current replica role.
+        /// </summary>
+        public ReplicaRole GetCurrentReplicaRole() => m_replicaRoleAccessor.Value?.Role ?? ReplicaRole.Unknown;
```

Comment:
I don't think this is useful, consumers generally don't access this class directly.